### PR TITLE
Fixed crashing when userSettings.xml was missing

### PR DESCRIPTION
**Pull request by idietmoran**
#37 
I hardcoded values in the event that the file does not exist. If values are not hard-coded in the window will never initialize and the build will fail. This is the easiest work around that works.

Please check the line at 55 

`SelectedDrives = new List<char>();`

I wasn't sure if this was the correct data type to use. Thanks.

### CODE REVIEW

**Comment by dfs-minded** [comment]
Hi @idietmoran! Thank you for taking this issue.
Is it possible to use values, which are defined in the UserSettings.xml file, included in the INdexerGUI project? Because now it is a duplication of information. Also, I would suggest not throwing exceptions, where it is a normal flow of the program and we can handle it.

**Comment by idietmoran** [comment]
@dfs-minded I agree with not throwing exceptions, but in this case it brings us to the catch statement and prevents the rest of the code from executing. I can change to an if else statement if you prefer, but I was just taking advantage of the `try / catch` statement that was there, otherwise the catch statement never gets logged in the first place on an error.

I can read from the file that is already there, and this should be a fairly straightforward change. I will fix when I get home later tonight.

**Comment by dfs-minded** [comment]
So I had a look at this issue more closely. What I would suggest is adding UserSettings as an embedded application resource and then, if the file is missing, create one from that resource.
```
if (!File.Exists(SerializationPath))
{
    File.WriteAllText("UserSettings.xml", Resources.UserAppSettings);
}
```
Adding usersettings as an embedded app resource:
adding usersettings as an embedded app resource

**Comment by idietmoran** [comment]
@dfs-minded ok, sounds good. I think we should rename it to something like `defaultSettings.xml` or something to avoid confusion.

**Comment by idietmoran** [comment]
Do you think it would be better to write the default user config, and the read the user config again, or just read the default user config and allow it to write on close.

**Comment by dfs-minded** [comment]
Both approaches are fine. But the less code in the result, the better. I suppose that the first approach will end up in fewer lines than the second one.

**Comment by idietmoran** [comment]
It's a trade off. I will defer to your judgement and fix this. Your solution is an easy fix.

File: IndexerGUI/IndexerGUI.csproj [hunk 1]
```diff
@@ -226,6 +226,9 @@
   </ItemGroup>
   <ItemGroup>
     <Resource Include="Icons\Indexer++Logo.png" />
+    <EmbeddedResource Include="defaultUserSettings.xml">
```
Comment (dfs-minded) [inline IndexerGUI/IndexerGUI.csproj]: Sorry, my fault that I didn't create the CONTRIBUTING file with a reference to code style guidelines we use. Have already fixed that. For resources name, we should use [upper case](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/naming-resources).

File: IndexerGUI/MainWindow.xaml.cs [hunk 2]
```diff
@@ -318,17 +318,25 @@ private void CheckUpdatesOnTimerTick(object sender, EventArgs eventArgs)
...
         {
             var settings = UserSettings.Instance;
+
+
             Height = settings.WndHeight;
             Width = settings.WndWidth;
             FiltersVisibility = settings.FiltersVisibility;
-
+            
+            ExcludeHiddenAndSystem = settings.ExcludeHiddenAndSystem;
```
Comment (dfs-minded) [inline IndexerGUI/MainWindow.xaml.cs]: Thanks for fixing this bug:)

File: IndexerGUI/MainWindow.xaml.cs [hunk 3]
```diff
@@ -318,17 +318,25 @@ private void CheckUpdatesOnTimerTick(object sender, EventArgs eventArgs)
...
+            ExcludeFolders = settings.ExcludeFolders;
+            
             var initialDirPath = CmdArgumentsParser.FilterDirPath;
             if (string.IsNullOrWhiteSpace(initialDirPath))
                 return;
 
             dirFilterEnabled = true;
             filtersVisibility = Visibility.Visible;
             searchDirPath = initialDirPath;
+
```
Comment (dfs-minded) [inline IndexerGUI/MainWindow.xaml.cs]: Remove these extra lines.

File: IndexerGUI/UserSettings.cs [hunk 4]
```diff
@@ -22,16 +23,16 @@ static UserSettings()
...
+        
         private UserSettings()
         {
             serializer = new DataContractSerializer(typeof (UserSettings));
             try
             {
                 if (!File.Exists(SerializationPath))
                 {
+                    File.WriteAllText(SerializationPath, Resources.defaultUserSettings);
                     Log.Instance.Error("User settings file does not exist in: " + SerializationPath);
```
Comment (dfs-minded) [inline IndexerGUI/UserSettings.cs]: Logging could be also removed now because it is not an error anymore.

**Comment by idietmoran** [comment]
yes, I will update them as soon as I get out of class.

**Comment by dfs-minded** [comment]
Thank you! As always, no rush:)

File: IndexerGUI/MainWindow.xaml.cs (hunk 3 above)
Comment (idietmoran) [reply IndexerGUI/MainWindow.xaml.cs]: Are you referring to the white space.

File: IndexerGUI/IndexerGUI.csproj [hunk 5]
```diff
@@ -225,7 +230,14 @@
...
   <ItemGroup>
+  <EmbeddedResource Include="Properties\Resources.resx">
+      <Generator>ResXFileCodeGenerator</Generator>
+      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
+    </EmbeddedResource>
     <Resource Include="Icons\Indexer++Logo.png" />
+    <EmbeddedResource Include="DefaultUserSettings.xml">
+      <SubType>Designer</SubType>
+    </EmbeddedResource>
     <Content Include="UserSettings.xml">
```
Comment (dfs-minded) [inline IndexerGUI/IndexerGUI.csproj]: This resource should be removed now.

File: IndexerGUI/Properties/Resources.Designer [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        ///  &lt;ExcludeFiles&gt;false&lt;/ExcludeFiles&gt;
+        ///  &lt;ExcludeFolders&gt;false&lt;/ExcludeFolders&gt;
+        ///  &lt;ExcludeHiddenAndSystem&gt;true&lt;/ExcludeHiddenAndSystem&gt;
+        ///  &lt;FiltersVisibility&gt;Collapsed&lt;/FiltersVisibility&gt;
+        ///  &lt;IsFiltersExpanded&gt;false&lt;/IsFiltersExpanded&gt;
+        ///  &lt;SelectedDrives xmlns:a=&quot;http://schemas.microsoft.com/2003/10/Serialization/Arrays&quot;&gt;&lt;/SelectedDrives&gt;
+        ///  &lt;WndHeight&gt;700&lt;/WndHeight&gt;
+        ///  &lt;WndWidth&gt;1000 [rest of string was truncated]&quot;;.
+        /// </summary>
+        internal static string defaultUserSettings {
```
Comment (dfs-minded) [inline IndexerGUI/Properties/Resources.Designer]: In the designer, it is still lower case. Better to recreate the resource from scratch.

**Comment by idietmoran** [comment]
aight, should be easy to clean it up in a new repo, sorry for the inconvenience.

**Comment by dfs-minded** [comment]
No problem, Git is not always intuitive:)

**Pull request closed without merging**